// @hyperledger/fabric-protos/lib/orderer/blockattestation_pb.js
// source: orderer/blockattestation.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = (function() {
  if (this) { return this; }
  if (typeof window !== 'undefined') { return window; }
  if (typeof global !== 'undefined') { return global; }
  if (typeof self !== 'undefined') { return self; }
  return Function('return this')();
}.call(null));

var common_common_pb = require('../common/common_pb.js');
goog.object.extend(proto, common_common_pb);
goog.exportSymbol('proto.orderer.BlockAttestation', null, global);
goog.exportSymbol('proto.orderer.BlockAttestationResponse', null, global);
goog.exportSymbol('proto.orderer.BlockAttestationResponse.TypeCase', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.orderer.BlockAttestation = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.orderer.BlockAttestation, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.orderer.BlockAttestation.displayName = 'proto.orderer.BlockAttestation';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.orderer.BlockAttestationResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, proto.orderer.BlockAttestationResponse.oneofGroups_);
};
goog.inherits(proto.orderer.BlockAttestationResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.orderer.BlockAttestationResponse.displayName = 'proto.orderer.BlockAttestationResponse';
}



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.orderer.BlockAttestation.prototype.toObject = function(opt_includeInstance) {
  return proto.orderer.BlockAttestation.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.orderer.BlockAttestation} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.orderer.BlockAttestation.toObject = function(includeInstance, msg) {
  var f, obj = {
    header: (f = msg.getHeader()) && common_common_pb.BlockHeader.toObject(includeInstance, f),
    metadata: (f = msg.getMetadata()) && common_common_pb.BlockMetadata.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.orderer.BlockAttestation}
 */
proto.orderer.BlockAttestation.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.orderer.BlockAttestation;
  return proto.orderer.BlockAttestation.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.orderer.BlockAttestation} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.orderer.BlockAttestation}
 */
proto.orderer.BlockAttestation.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = new common_common_pb.BlockHeader;
      reader.readMessage(value,common_common_pb.BlockHeader.deserializeBinaryFromReader);
      msg.setHeader(value);
      break;
    case 2:
      var value = new common_common_pb.BlockMetadata;
      reader.readMessage(value,common_common_pb.BlockMetadata.deserializeBinaryFromReader);
      msg.setMetadata(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.orderer.BlockAttestation.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.orderer.BlockAttestation.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.orderer.BlockAttestation} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.orderer.BlockAttestation.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getHeader();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      common_common_pb.BlockHeader.serializeBinaryToWriter
    );
  }
  f = message.getMetadata();
  if (f != null) {
    writer.writeMessage(
      2,
      f,
      common_common_pb.BlockMetadata.serializeBinaryToWriter
    );
  }
};


/**
 * optional common.BlockHeader header = 1;
 * @return {?proto.common.BlockHeader}
 */
proto.orderer.BlockAttestation.prototype.getHeader = function() {
  return /** @type{?proto.common.BlockHeader} */ (
    jspb.Message.getWrapperField(this, common_common_pb.BlockHeader, 1));
};


/**
 * @param {?proto.common.BlockHeader|undefined} value
 * @return {!proto.orderer.BlockAttestation} returns this
*/
proto.orderer.BlockAttestation.prototype.setHeader = function(value) {
  return jspb.Message.setWrapperField(this, 1, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.orderer.BlockAttestation} returns this
 */
proto.orderer.BlockAttestation.prototype.clearHeader = function() {
  return this.setHeader(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.orderer.BlockAttestation.prototype.hasHeader = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional common.BlockMetadata metadata = 2;
 * @return {?proto.common.BlockMetadata}
 */
proto.orderer.BlockAttestation.prototype.getMetadata = function() {
  return /** @type{?proto.common.BlockMetadata} */ (
    jspb.Message.getWrapperField(this, common_common_pb.BlockMetadata, 2));
};


/**
 * @param {?proto.common.BlockMetadata|undefined} value
 * @return {!proto.orderer.BlockAttestation} returns this
*/
proto.orderer.BlockAttestation.prototype.setMetadata = function(value) {
  return jspb.Message.setWrapperField(this, 2, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.orderer.BlockAttestation} returns this
 */
proto.orderer.BlockAttestation.prototype.clearMetadata = function() {
  return this.setMetadata(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.orderer.BlockAttestation.prototype.hasMetadata = function() {
  return jspb.Message.getField(this, 2) != null;
};



/**
 * Oneof group definitions for this message. Each group defines the field
 * numbers belonging to that group. When of these fields' value is set, all
 * other fields in the group are cleared. During deserialization, if multiple
 * fields are encountered for a group, only the last value seen will be kept.
 * @private {!Array<!Array<number>>}
 * @const
 */
proto.orderer.BlockAttestationResponse.oneofGroups_ = [[1,2]];

/**
 * @enum {number}
 */
proto.orderer.BlockAttestationResponse.TypeCase = {
  TYPE_NOT_SET: 0,
  STATUS: 1,
  BLOCK_ATTESTATION: 2
};

/**
 * @return {proto.orderer.BlockAttestationResponse.TypeCase}
 */
proto.orderer.BlockAttestationResponse.prototype.getTypeCase = function() {
  return /** @type {proto.orderer.BlockAttestationResponse.TypeCase} */(jspb.Message.computeOneofCase(this, proto.orderer.BlockAttestationResponse.oneofGroups_[0]));
};



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.orderer.BlockAttestationResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.orderer.BlockAttestationResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.orderer.BlockAttestationResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.orderer.BlockAttestationResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    status: jspb.Message.getFieldWithDefault(msg, 1, 0),
    blockAttestation: (f = msg.getBlockAttestation()) && proto.orderer.BlockAttestation.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.orderer.BlockAttestationResponse}
 */
proto.orderer.BlockAttestationResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.orderer.BlockAttestationResponse;
  return proto.orderer.BlockAttestationResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.orderer.BlockAttestationResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.orderer.BlockAttestationResponse}
 */
proto.orderer.BlockAttestationResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {!proto.common.Status} */ (reader.readEnum());
      msg.setStatus(value);
      break;
    case 2:
      var value = new proto.orderer.BlockAttestation;
      reader.readMessage(value,proto.orderer.BlockAttestation.deserializeBinaryFromReader);
      msg.setBlockAttestation(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.orderer.BlockAttestationResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.orderer.BlockAttestationResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.orderer.BlockAttestationResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.orderer.BlockAttestationResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = /** @type {!proto.common.Status} */ (jspb.Message.getField(message, 1));
  if (f != null) {
    writer.writeEnum(
      1,
      f
    );
  }
  f = message.getBlockAttestation();
  if (f != null) {
    writer.writeMessage(
      2,
      f,
      proto.orderer.BlockAttestation.serializeBinaryToWriter
    );
  }
};


/**
 * optional common.Status status = 1;
 * @return {!proto.common.Status}
 */
proto.orderer.BlockAttestationResponse.prototype.getStatus = function() {
  return /** @type {!proto.common.Status} */ (jspb.Message.getFieldWithDefault(this, 1, 0));
};


/**
 * @param {!proto.common.Status} value
 * @return {!proto.orderer.BlockAttestationResponse} returns this
 */
proto.orderer.BlockAttestationResponse.prototype.setStatus = function(value) {
  return jspb.Message.setOneofField(this, 1, proto.orderer.BlockAttestationResponse.oneofGroups_[0], value);
};


/**
 * Clears the field making it undefined.
 * @return {!proto.orderer.BlockAttestationResponse} returns this
 */
proto.orderer.BlockAttestationResponse.prototype.clearStatus = function() {
  return jspb.Message.setOneofField(this, 1, proto.orderer.BlockAttestationResponse.oneofGroups_[0], undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.orderer.BlockAttestationResponse.prototype.hasStatus = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional BlockAttestation block_attestation = 2;
 * @return {?proto.orderer.BlockAttestation}
 */
proto.orderer.BlockAttestationResponse.prototype.getBlockAttestation = function() {
  return /** @type{?proto.orderer.BlockAttestation} */ (
    jspb.Message.getWrapperField(this, proto.orderer.BlockAttestation, 2));
};


/**
 * @param {?proto.orderer.BlockAttestation|undefined} value
 * @return {!proto.orderer.BlockAttestationResponse} returns this
*/
proto.orderer.BlockAttestationResponse.prototype.setBlockAttestation = function(value) {
  return jspb.Message.setOneofWrapperField(this, 2, proto.orderer.BlockAttestationResponse.oneofGroups_[0], value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.orderer.BlockAttestationResponse} returns this
 */
proto.orderer.BlockAttestationResponse.prototype.clearBlockAttestation = function() {
  return this.setBlockAttestation(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.orderer.BlockAttestationResponse.prototype.hasBlockAttestation = function() {
  return jspb.Message.getField(this, 2) != null;
};


goog.object.extend(exports, proto.orderer);
