// @hyperledger/fabric-gateway/dist/blockeventsbuilder.js
"use strict";
/*
 * Copyright 2022 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _BlockEventsBuilder_options, _BlockEventsBuilder_envelopeBuilder, _FilteredBlockEventsBuilder_options, _FilteredBlockEventsBuilder_envelopeBuilder, _BlockAndPrivateDataEventsBuilder_options, _BlockAndPrivateDataEventsBuilder_envelopeBuilder, _BlockEventsEnvelopeBuilder_instances, _BlockEventsEnvelopeBuilder_options, _BlockEventsEnvelopeBuilder_eventsBuilder, _BlockEventsEnvelopeBuilder_newPayload, _BlockEventsEnvelopeBuilder_newHeader, _BlockEventsEnvelopeBuilder_newChannelHeader, _BlockEventsEnvelopeBuilder_newSignatureHeader, _BlockEventsEnvelopeBuilder_newSeekInfo;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockAndPrivateDataEventsBuilder = exports.FilteredBlockEventsBuilder = exports.BlockEventsBuilder = void 0;
const fabric_protos_1 = require("@hyperledger/fabric-protos");
const timestamp_pb_1 = require("google-protobuf/google/protobuf/timestamp_pb");
const blockeventsrequest_1 = require("./blockeventsrequest");
const eventsbuilder_1 = require("./eventsbuilder");
function seekLargestBlockNumber() {
    const largestBlockNumber = new fabric_protos_1.orderer.SeekSpecified();
    largestBlockNumber.setNumber(Number.MAX_SAFE_INTEGER);
    const result = new fabric_protos_1.orderer.SeekPosition();
    result.setSpecified(largestBlockNumber);
    return result;
}
class BlockEventsBuilder {
    constructor(options) {
        _BlockEventsBuilder_options.set(this, void 0);
        _BlockEventsBuilder_envelopeBuilder.set(this, void 0);
        __classPrivateFieldSet(this, _BlockEventsBuilder_options, options, "f");
        __classPrivateFieldSet(this, _BlockEventsBuilder_envelopeBuilder, new BlockEventsEnvelopeBuilder(options), "f");
    }
    build() {
        const request = __classPrivateFieldGet(this, _BlockEventsBuilder_envelopeBuilder, "f").build();
        return new blockeventsrequest_1.BlockEventsRequestImpl({
            client: __classPrivateFieldGet(this, _BlockEventsBuilder_options, "f").client,
            request,
            signingIdentity: __classPrivateFieldGet(this, _BlockEventsBuilder_options, "f").signingIdentity,
        });
    }
}
exports.BlockEventsBuilder = BlockEventsBuilder;
_BlockEventsBuilder_options = new WeakMap(), _BlockEventsBuilder_envelopeBuilder = new WeakMap();
class FilteredBlockEventsBuilder {
    constructor(options) {
        _FilteredBlockEventsBuilder_options.set(this, void 0);
        _FilteredBlockEventsBuilder_envelopeBuilder.set(this, void 0);
        __classPrivateFieldSet(this, _FilteredBlockEventsBuilder_options, options, "f");
        __classPrivateFieldSet(this, _FilteredBlockEventsBuilder_envelopeBuilder, new BlockEventsEnvelopeBuilder(options), "f");
    }
    build() {
        const request = __classPrivateFieldGet(this, _FilteredBlockEventsBuilder_envelopeBuilder, "f").build();
        return new blockeventsrequest_1.FilteredBlockEventsRequestImpl({
            client: __classPrivateFieldGet(this, _FilteredBlockEventsBuilder_options, "f").client,
            request,
            signingIdentity: __classPrivateFieldGet(this, _FilteredBlockEventsBuilder_options, "f").signingIdentity,
        });
    }
}
exports.FilteredBlockEventsBuilder = FilteredBlockEventsBuilder;
_FilteredBlockEventsBuilder_options = new WeakMap(), _FilteredBlockEventsBuilder_envelopeBuilder = new WeakMap();
class BlockAndPrivateDataEventsBuilder {
    constructor(options) {
        _BlockAndPrivateDataEventsBuilder_options.set(this, void 0);
        _BlockAndPrivateDataEventsBuilder_envelopeBuilder.set(this, void 0);
        __classPrivateFieldSet(this, _BlockAndPrivateDataEventsBuilder_options, options, "f");
        __classPrivateFieldSet(this, _BlockAndPrivateDataEventsBuilder_envelopeBuilder, new BlockEventsEnvelopeBuilder(options), "f");
    }
    build() {
        const request = __classPrivateFieldGet(this, _BlockAndPrivateDataEventsBuilder_envelopeBuilder, "f").build();
        return new blockeventsrequest_1.BlockAndPrivateDataEventsRequestImpl({
            client: __classPrivateFieldGet(this, _BlockAndPrivateDataEventsBuilder_options, "f").client,
            request,
            signingIdentity: __classPrivateFieldGet(this, _BlockAndPrivateDataEventsBuilder_options, "f").signingIdentity,
        });
    }
}
exports.BlockAndPrivateDataEventsBuilder = BlockAndPrivateDataEventsBuilder;
_BlockAndPrivateDataEventsBuilder_options = new WeakMap(), _BlockAndPrivateDataEventsBuilder_envelopeBuilder = new WeakMap();
class BlockEventsEnvelopeBuilder {
    constructor(options) {
        _BlockEventsEnvelopeBuilder_instances.add(this);
        _BlockEventsEnvelopeBuilder_options.set(this, void 0);
        _BlockEventsEnvelopeBuilder_eventsBuilder.set(this, void 0);
        __classPrivateFieldSet(this, _BlockEventsEnvelopeBuilder_options, options, "f");
        __classPrivateFieldSet(this, _BlockEventsEnvelopeBuilder_eventsBuilder, new eventsbuilder_1.EventsBuilder(options), "f");
    }
    build() {
        const result = new fabric_protos_1.common.Envelope();
        result.setPayload(__classPrivateFieldGet(this, _BlockEventsEnvelopeBuilder_instances, "m", _BlockEventsEnvelopeBuilder_newPayload).call(this).serializeBinary());
        return result;
    }
}
_BlockEventsEnvelopeBuilder_options = new WeakMap(), _BlockEventsEnvelopeBuilder_eventsBuilder = new WeakMap(), _BlockEventsEnvelopeBuilder_instances = new WeakSet(), _BlockEventsEnvelopeBuilder_newPayload = function _BlockEventsEnvelopeBuilder_newPayload() {
    const result = new fabric_protos_1.common.Payload();
    result.setHeader(__classPrivateFieldGet(this, _BlockEventsEnvelopeBuilder_instances, "m", _BlockEventsEnvelopeBuilder_newHeader).call(this));
    result.setData(__classPrivateFieldGet(this, _BlockEventsEnvelopeBuilder_instances, "m", _BlockEventsEnvelopeBuilder_newSeekInfo).call(this).serializeBinary());
    return result;
}, _BlockEventsEnvelopeBuilder_newHeader = function _BlockEventsEnvelopeBuilder_newHeader() {
    const result = new fabric_protos_1.common.Header();
    result.setChannelHeader(__classPrivateFieldGet(this, _BlockEventsEnvelopeBuilder_instances, "m", _BlockEventsEnvelopeBuilder_newChannelHeader).call(this).serializeBinary());
    result.setSignatureHeader(__classPrivateFieldGet(this, _BlockEventsEnvelopeBuilder_instances, "m", _BlockEventsEnvelopeBuilder_newSignatureHeader).call(this).serializeBinary());
    return result;
}, _BlockEventsEnvelopeBuilder_newChannelHeader = function _BlockEventsEnvelopeBuilder_newChannelHeader() {
    const result = new fabric_protos_1.common.ChannelHeader();
    result.setChannelId(__classPrivateFieldGet(this, _BlockEventsEnvelopeBuilder_options, "f").channelName);
    result.setEpoch(0);
    result.setTimestamp(timestamp_pb_1.Timestamp.fromDate(new Date()));
    result.setType(fabric_protos_1.common.HeaderType.DELIVER_SEEK_INFO);
    return result;
}, _BlockEventsEnvelopeBuilder_newSignatureHeader = function _BlockEventsEnvelopeBuilder_newSignatureHeader() {
    const result = new fabric_protos_1.common.SignatureHeader();
    result.setCreator(__classPrivateFieldGet(this, _BlockEventsEnvelopeBuilder_options, "f").signingIdentity.getCreator());
    return result;
}, _BlockEventsEnvelopeBuilder_newSeekInfo = function _BlockEventsEnvelopeBuilder_newSeekInfo() {
    const result = new fabric_protos_1.orderer.SeekInfo();
    result.setStart(__classPrivateFieldGet(this, _BlockEventsEnvelopeBuilder_eventsBuilder, "f").getStartPosition());
    result.setStop(seekLargestBlockNumber());
    return result;
};
