// @hyperledger/fabric-gateway/dist/chaincodeevent.js
"use strict";
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.newChaincodeEvents = void 0;
function newChaincodeEvents(responses) {
    return {
        async *[Symbol.asyncIterator]() {
            for await (const response of responses) {
                const blockNumber = BigInt(response.getBlockNumber() ?? 0);
                const events = response.getEventsList() || [];
                for (const event of events) {
                    yield newChaincodeEvent(blockNumber, event);
                }
            }
        },
        close: () => {
            responses.close();
        },
    };
}
exports.newChaincodeEvents = newChaincodeEvents;
function newChaincodeEvent(blockNumber, event) {
    return {
        blockNumber,
        chaincodeName: event.getChaincodeId() ?? '',
        eventName: event.getEventName() ?? '',
        transactionId: event.getTxId() ?? '',
        payload: event.getPayload_asU8() || new Uint8Array(),
    };
}
