// @hyperledger/fabric-gateway/dist/proposal.js
"use strict";
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ProposalImpl_instances, _ProposalImpl_client, _ProposalImpl_signingIdentity, _ProposalImpl_channelName, _ProposalImpl_proposedTransaction, _ProposalImpl_proposal, _ProposalImpl_sign, _ProposalImpl_isSigned, _ProposalImpl_newEvaluateRequest, _ProposalImpl_newEndorseRequest, _ProposalImpl_newPreparedTransaction;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProposalImpl = void 0;
const fabric_protos_1 = require("@hyperledger/fabric-protos");
const gateway_1 = require("./gateway");
const transaction_1 = require("./transaction");
class ProposalImpl {
    constructor(options) {
        _ProposalImpl_instances.add(this);
        _ProposalImpl_client.set(this, void 0);
        _ProposalImpl_signingIdentity.set(this, void 0);
        _ProposalImpl_channelName.set(this, void 0);
        _ProposalImpl_proposedTransaction.set(this, void 0);
        _ProposalImpl_proposal.set(this, void 0);
        __classPrivateFieldSet(this, _ProposalImpl_client, options.client, "f");
        __classPrivateFieldSet(this, _ProposalImpl_signingIdentity, options.signingIdentity, "f");
        __classPrivateFieldSet(this, _ProposalImpl_channelName, options.channelName, "f");
        __classPrivateFieldSet(this, _ProposalImpl_proposedTransaction, options.proposedTransaction, "f");
        __classPrivateFieldSet(this, _ProposalImpl_proposal, (0, gateway_1.assertDefined)(options.proposedTransaction.getProposal(), 'Missing signed proposal'), "f");
    }
    getBytes() {
        return __classPrivateFieldGet(this, _ProposalImpl_proposedTransaction, "f").serializeBinary();
    }
    getDigest() {
        const bytes = __classPrivateFieldGet(this, _ProposalImpl_proposal, "f").getProposalBytes_asU8();
        return __classPrivateFieldGet(this, _ProposalImpl_signingIdentity, "f").hash(bytes);
    }
    getTransactionId() {
        return __classPrivateFieldGet(this, _ProposalImpl_proposedTransaction, "f").getTransactionId();
    }
    async evaluate(options) {
        await __classPrivateFieldGet(this, _ProposalImpl_instances, "m", _ProposalImpl_sign).call(this);
        const evaluateResponse = await __classPrivateFieldGet(this, _ProposalImpl_client, "f").evaluate(__classPrivateFieldGet(this, _ProposalImpl_instances, "m", _ProposalImpl_newEvaluateRequest).call(this), options);
        const result = evaluateResponse.getResult();
        return result?.getPayload_asU8() ?? new Uint8Array(0);
    }
    async endorse(options) {
        await __classPrivateFieldGet(this, _ProposalImpl_instances, "m", _ProposalImpl_sign).call(this);
        const endorseResponse = await __classPrivateFieldGet(this, _ProposalImpl_client, "f").endorse(__classPrivateFieldGet(this, _ProposalImpl_instances, "m", _ProposalImpl_newEndorseRequest).call(this), options);
        const txEnvelope = (0, gateway_1.assertDefined)(endorseResponse.getPreparedTransaction(), 'Missing transaction envelope');
        return new transaction_1.TransactionImpl({
            client: __classPrivateFieldGet(this, _ProposalImpl_client, "f"),
            signingIdentity: __classPrivateFieldGet(this, _ProposalImpl_signingIdentity, "f"),
            preparedTransaction: __classPrivateFieldGet(this, _ProposalImpl_instances, "m", _ProposalImpl_newPreparedTransaction).call(this, txEnvelope)
        });
    }
    setSignature(signature) {
        __classPrivateFieldGet(this, _ProposalImpl_proposal, "f").setSignature(signature);
    }
}
exports.ProposalImpl = ProposalImpl;
_ProposalImpl_client = new WeakMap(), _ProposalImpl_signingIdentity = new WeakMap(), _ProposalImpl_channelName = new WeakMap(), _ProposalImpl_proposedTransaction = new WeakMap(), _ProposalImpl_proposal = new WeakMap(), _ProposalImpl_instances = new WeakSet(), _ProposalImpl_sign = async function _ProposalImpl_sign() {
    if (__classPrivateFieldGet(this, _ProposalImpl_instances, "m", _ProposalImpl_isSigned).call(this)) {
        return;
    }
    const signature = await __classPrivateFieldGet(this, _ProposalImpl_signingIdentity, "f").sign(this.getDigest());
    this.setSignature(signature);
}, _ProposalImpl_isSigned = function _ProposalImpl_isSigned() {
    const signatureLength = __classPrivateFieldGet(this, _ProposalImpl_proposal, "f").getSignature_asU8().length;
    return signatureLength > 0;
}, _ProposalImpl_newEvaluateRequest = function _ProposalImpl_newEvaluateRequest() {
    const result = new fabric_protos_1.gateway.EvaluateRequest();
    result.setTransactionId(__classPrivateFieldGet(this, _ProposalImpl_proposedTransaction, "f").getTransactionId());
    result.setChannelId(__classPrivateFieldGet(this, _ProposalImpl_channelName, "f"));
    result.setProposedTransaction(__classPrivateFieldGet(this, _ProposalImpl_proposal, "f"));
    result.setTargetOrganizationsList(__classPrivateFieldGet(this, _ProposalImpl_proposedTransaction, "f").getEndorsingOrganizationsList());
    return result;
}, _ProposalImpl_newEndorseRequest = function _ProposalImpl_newEndorseRequest() {
    const result = new fabric_protos_1.gateway.EndorseRequest();
    result.setTransactionId(__classPrivateFieldGet(this, _ProposalImpl_proposedTransaction, "f").getTransactionId());
    result.setChannelId(__classPrivateFieldGet(this, _ProposalImpl_channelName, "f"));
    result.setProposedTransaction(__classPrivateFieldGet(this, _ProposalImpl_proposal, "f"));
    result.setEndorsingOrganizationsList(__classPrivateFieldGet(this, _ProposalImpl_proposedTransaction, "f").getEndorsingOrganizationsList());
    return result;
}, _ProposalImpl_newPreparedTransaction = function _ProposalImpl_newPreparedTransaction(envelope) {
    const result = new fabric_protos_1.gateway.PreparedTransaction();
    result.setEnvelope(envelope);
    result.setTransactionId(__classPrivateFieldGet(this, _ProposalImpl_proposedTransaction, "f").getTransactionId());
    return result;
};
