// @hyperledger/fabric-gateway/dist/proposalbuilder.js
"use strict";
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ProposalBuilder_instances, _ProposalBuilder_options, _ProposalBuilder_transactionContext, _ProposalBuilder_newProposedTransaction, _ProposalBuilder_newSignedProposal, _ProposalBuilder_newProposal, _ProposalBuilder_newHeader, _ProposalBuilder_newChannelHeader, _ProposalBuilder_newChaincodeHeaderExtension, _ProposalBuilder_newChaincodeID, _ProposalBuilder_newChaincodeProposalPayload, _ProposalBuilder_newChaincodeInvocationSpec, _ProposalBuilder_newChaincodeSpec, _ProposalBuilder_newChaincodeInput, _ProposalBuilder_getArgsAsBytes, _ProposalBuilder_getTransientData;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProposalBuilder = void 0;
const fabric_protos_1 = require("@hyperledger/fabric-protos");
const timestamp_pb_1 = require("google-protobuf/google/protobuf/timestamp_pb");
const proposal_1 = require("./proposal");
const transactioncontext_1 = require("./transactioncontext");
class ProposalBuilder {
    constructor(options) {
        _ProposalBuilder_instances.add(this);
        _ProposalBuilder_options.set(this, void 0);
        _ProposalBuilder_transactionContext.set(this, void 0);
        __classPrivateFieldSet(this, _ProposalBuilder_options, options, "f");
        __classPrivateFieldSet(this, _ProposalBuilder_transactionContext, new transactioncontext_1.TransactionContext(options.signingIdentity), "f");
    }
    build() {
        return new proposal_1.ProposalImpl({
            client: __classPrivateFieldGet(this, _ProposalBuilder_options, "f").client,
            signingIdentity: __classPrivateFieldGet(this, _ProposalBuilder_options, "f").signingIdentity,
            channelName: __classPrivateFieldGet(this, _ProposalBuilder_options, "f").channelName,
            proposedTransaction: __classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newProposedTransaction).call(this),
        });
    }
}
exports.ProposalBuilder = ProposalBuilder;
_ProposalBuilder_options = new WeakMap(), _ProposalBuilder_transactionContext = new WeakMap(), _ProposalBuilder_instances = new WeakSet(), _ProposalBuilder_newProposedTransaction = function _ProposalBuilder_newProposedTransaction() {
    const result = new fabric_protos_1.gateway.ProposedTransaction();
    result.setProposal(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newSignedProposal).call(this));
    result.setTransactionId(__classPrivateFieldGet(this, _ProposalBuilder_transactionContext, "f").getTransactionId());
    if (__classPrivateFieldGet(this, _ProposalBuilder_options, "f").endorsingOrganizations) {
        result.setEndorsingOrganizationsList(__classPrivateFieldGet(this, _ProposalBuilder_options, "f").endorsingOrganizations);
    }
    return result;
}, _ProposalBuilder_newSignedProposal = function _ProposalBuilder_newSignedProposal() {
    const result = new fabric_protos_1.peer.SignedProposal();
    result.setProposalBytes(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newProposal).call(this).serializeBinary());
    return result;
}, _ProposalBuilder_newProposal = function _ProposalBuilder_newProposal() {
    const result = new fabric_protos_1.peer.Proposal();
    result.setHeader(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newHeader).call(this).serializeBinary());
    result.setPayload(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newChaincodeProposalPayload).call(this).serializeBinary());
    return result;
}, _ProposalBuilder_newHeader = function _ProposalBuilder_newHeader() {
    const result = new fabric_protos_1.common.Header();
    result.setChannelHeader(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newChannelHeader).call(this).serializeBinary());
    result.setSignatureHeader(__classPrivateFieldGet(this, _ProposalBuilder_transactionContext, "f").getSignatureHeader().serializeBinary());
    return result;
}, _ProposalBuilder_newChannelHeader = function _ProposalBuilder_newChannelHeader() {
    const result = new fabric_protos_1.common.ChannelHeader();
    result.setType(fabric_protos_1.common.HeaderType.ENDORSER_TRANSACTION);
    result.setTxId(__classPrivateFieldGet(this, _ProposalBuilder_transactionContext, "f").getTransactionId());
    result.setTimestamp(timestamp_pb_1.Timestamp.fromDate(new Date()));
    result.setChannelId(__classPrivateFieldGet(this, _ProposalBuilder_options, "f").channelName);
    result.setExtension$(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newChaincodeHeaderExtension).call(this).serializeBinary());
    result.setEpoch(0);
    return result;
}, _ProposalBuilder_newChaincodeHeaderExtension = function _ProposalBuilder_newChaincodeHeaderExtension() {
    const result = new fabric_protos_1.peer.ChaincodeHeaderExtension();
    result.setChaincodeId(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newChaincodeID).call(this));
    return result;
}, _ProposalBuilder_newChaincodeID = function _ProposalBuilder_newChaincodeID() {
    const result = new fabric_protos_1.peer.ChaincodeID();
    result.setName(__classPrivateFieldGet(this, _ProposalBuilder_options, "f").chaincodeName);
    return result;
}, _ProposalBuilder_newChaincodeProposalPayload = function _ProposalBuilder_newChaincodeProposalPayload() {
    const result = new fabric_protos_1.peer.ChaincodeProposalPayload();
    result.setInput(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newChaincodeInvocationSpec).call(this).serializeBinary());
    const transientMap = result.getTransientmapMap();
    for (const [key, value] of Object.entries(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_getTransientData).call(this))) {
        transientMap.set(key, value);
    }
    return result;
}, _ProposalBuilder_newChaincodeInvocationSpec = function _ProposalBuilder_newChaincodeInvocationSpec() {
    const result = new fabric_protos_1.peer.ChaincodeInvocationSpec();
    result.setChaincodeSpec(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newChaincodeSpec).call(this));
    return result;
}, _ProposalBuilder_newChaincodeSpec = function _ProposalBuilder_newChaincodeSpec() {
    const result = new fabric_protos_1.peer.ChaincodeSpec();
    result.setType(fabric_protos_1.peer.ChaincodeSpec.Type.NODE);
    result.setChaincodeId(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newChaincodeID).call(this));
    result.setInput(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_newChaincodeInput).call(this));
    return result;
}, _ProposalBuilder_newChaincodeInput = function _ProposalBuilder_newChaincodeInput() {
    const result = new fabric_protos_1.peer.ChaincodeInput();
    result.setArgsList(__classPrivateFieldGet(this, _ProposalBuilder_instances, "m", _ProposalBuilder_getArgsAsBytes).call(this));
    return result;
}, _ProposalBuilder_getArgsAsBytes = function _ProposalBuilder_getArgsAsBytes() {
    return Array.of(__classPrivateFieldGet(this, _ProposalBuilder_options, "f").transactionName, ...(__classPrivateFieldGet(this, _ProposalBuilder_options, "f").arguments ?? []))
        .map(asBytes);
}, _ProposalBuilder_getTransientData = function _ProposalBuilder_getTransientData() {
    const result = {};
    for (const [key, value] of Object.entries(__classPrivateFieldGet(this, _ProposalBuilder_options, "f").transientData ?? {})) {
        result[key] = asBytes(value);
    }
    return result;
};
function asBytes(value) {
    return typeof value === 'string' ? Buffer.from(value) : value;
}
