// @hyperledger/fabric-protos/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
exports.__esModule = true;
exports.google = exports.ledger = exports.orderer = exports.peer = exports.transientstore = exports.discovery = exports.msp = exports.common = exports.gateway = exports.gossip = void 0;
exports.gossip = __importStar(require("./gossip"));
exports.gateway = __importStar(require("./gateway"));
exports.common = __importStar(require("./common"));
exports.msp = __importStar(require("./msp"));
exports.discovery = __importStar(require("./discovery"));
exports.transientstore = __importStar(require("./transientstore"));
exports.peer = __importStar(require("./peer"));
exports.orderer = __importStar(require("./orderer"));
exports.ledger = __importStar(require("./ledger"));
exports.google = __importStar(require("./google"));
